registry/storage/shares.go

```diff
@@ -315,95 +318,88 @@ func (s *sharesStorage) storageShareToSpecShare(share *storageShare) (*types.SSV
 	return specShare, nil
 }
 
+var errShareNotFound = errors.New("share not found")
+
 func (s *sharesStorage) Delete(rw basedb.ReadWriter, pubKey []byte) error {
-	s.mu.Lock()
-	defer s.mu.Unlock()
+	s.dbmu.Lock()
+	defer s.dbmu.Unlock()
 
-	// Delete the share from the database
-	if err := s.db.Using(rw).Delete(s.prefix, s.storageKey(pubKey)); err != nil {
-		return err
-	}
+	err := func() error {
+		s.mu.Lock()
+		defer s.mu.Unlock()
 
-	share := s.shares[hex.EncodeToString(pubKey)]
-	if share == nil {
-		return nil
-	}
+		share, found := s.shares[hex.EncodeToString(pubKey)]
+		if !found {
+			return errShareNotFound
+		}
 
-	// Remove the share from local storage map
-	delete(s.shares, hex.EncodeToString(pubKey))
+		// Remove the share from local storage map
+		delete(s.shares, hex.EncodeToString(pubKey))
 
-	// Remove the share from the validator store. This method will handle its own locking.
-	if err := s.validatorStore.handleShareRemoved(share); err != nil {
+		// Remove the share from the validator store. This method will handle its own locking.
+		return s.validatorStore.handleShareRemoved(share)
+	}()
+	if errors.Is(err, errShareNotFound) {
+		return nil
+	}
+	if err != nil {
 		return err
 	}
 
-	return nil
+	// Delete the share from the database
+	return s.db.Using(rw).Delete(s.prefix, s.storageKey(pubKey))
 }
 
 // UpdateValidatorsMetadata updates the metadata of the given validator
 func (s *sharesStorage) UpdateValidatorsMetadata(data map[spectypes.ValidatorPK]*beaconprotocol.ValidatorMetadata) error {
 	var shares []*types.SSVShare
 
-	func() {
-		s.mu.RLock()
-		defer s.mu.RUnlock()
+	s.dbmu.Lock()
+	defer s.dbmu.Unlock()
+
+	err := func() error {
+		s.mu.Lock()
+		defer s.mu.Unlock()
```

Comment:
Here we can use `RLock ?

Reply:
re-added `s.shares[key] = share` in the same scope